sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,267 @@
+
+    n_iter_ : int
+        The amount of iterations the classifier used during fitting.
+
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> rng = np.random.RandomState(42)
+    >>> iris = datasets.load_iris()
+    >>> random_unlabeled_points = rng.rand(iris.target.shape[0]) < 0.3
+    >>> iris.target[random_unlabeled_points] = -1
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> self_training_model.fit(iris.data, iris.target)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    _estimator_type = "classifier"
+
+    def __init__(self,
+                 base_classifier,
+                 threshold=0.75,
+                 max_iter=10):
+        self.base_classifier = base_classifier
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to dataset, using the
+        base_classifier passed. The fitted base_classifier is stored as
+        ``self.base_classifier_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples,)
+            array representing the labels. Unlabeled samples should have the
+            label -1.
+
+        Returns
+        -------
+        self : object
+            returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _validate_estimator(self.base_classifier)
+        self.base_classifier_ = clone(self.base_classifier)
+
+        if self.max_iter is not None and self.max_iter < 0:
+            msg = "max_iter must be >= 0 or None, got {}".format(self.max_iter)
+            raise ValueError(msg)
+
+        if not 0 <= self.threshold < 1:
+            msg = "threshold must be in [0,1), got {}".format(self.threshold)
+            raise ValueError(msg)
+
+        if y.dtype.kind == 'U' or y.dtype.kind == 'S':
+            has_label = y != '-1'
+        else:
+            has_label = y != -1
+
+        if np.all(has_label):
+            warnings.warn(RuntimeWarning("y contains no unlabeled samples"))
+
+        self.y_labels_ = np.copy(y)
+        self.y_labeled_iter_ = np.full_like(y, -1)
+        self.y_labeled_iter_[has_label] = 0
+
+        self.n_iter_ = 0
+        while not np.all(has_label) and (self.max_iter is None or
```

Comment:
But you also need another stopping criterion of `len(new_labels_idx) == 0` below.

You may also consider writing this as a for loop:

```
max_iter = self.max_iter
if max_iter is None:
    max_iter = len(y)  # this should be more than sufficient
for self.n_iter_ in range(1, max_iter + 1):
```

Reply:
Again, I'm not sure how this would work with string inputs. If we say that string arrays should have unlabeled samples as -1 (an integer), we need the y array to have dtype=object, but passing such an array to most classifiers (for example SVC) will cause an error when those classifiers call `check_classification_targets`.

Reply:
I will rewrite the fit so that we first convert string arrays to int arrays at the start of the fit, and then we can use the minimum = -1 condition. It should also perform slightly better.